### PR TITLE
New Console Logger Test and Flush before Eval

### PR DESCRIPTION
**Pull request by rishab-partha**
# What does this PR do?

Changes the behavior of the console logger to empty before eval instead of after each epoch. Added unit test `test_console_logger_overlapping` to test this logic by monitoring the optimizer.

# What issue(s) does this change relate to?

[CO-2152](https://mosaicml.atlassian.net/browse/CO-2152
)

[CO-2152]: https://mosaicml.atlassian.net/browse/CO-2152?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: composer/loggers/console_logger.py [hunk 1]
```diff
@@ -134,6 +130,9 @@ def predict_start(self, state: State, logger: Logger) -> None:
             self._log_hparams_to_console()
 
     def eval_start(self, state: State, logger: Logger) -> None:
+        self.logged_metrics = {}
+        # Clear logged metrics before eval so they don't get logged
+        # subsequently.
```
Comment (mvpatel2000) [inline composer/loggers/console_logger.py:135]: nit: let's one line as much as possible
```suggestion
        self.logged_metrics = {}  # Clear logged metrics before eval so they don't get logged
```

**Review by mvpatel2000 — COMMENTED** [review]
Looks like this PR is stacked on top of the previous one, can you move commits to a new fork? otherwise we can wait to resolve the other PR first

File: composer/callbacks/mlperf.py [hunk 2]
```diff
@@ -87,7 +87,7 @@ class MLPerfCallback(Callback):
                 target='0.759',
             )
 
-    During training, the metric found in ``state.eval_metrics[metric_label][metric_name]``
+    During training, the metric found in ``state.eval_metrics[metric_name][metric_label]``
```
Comment (dakinggg) [inline composer/callbacks/mlperf.py:90]:
```suggestion
    During training, the metric found in ``state.eval_metrics[evaluator_label][metric_name]``
```
^is this correct?

File: composer/trainer/trainer.py [hunk 3]
```diff
@@ -469,12 +469,13 @@ class Trainer:
             Setting this to ``True`` will force schedulers to be stepped every batch,
             while ``False`` means schedulers stepped every epoch. ``None`` indicates the default behavior.
             (default: ``None``)
-        eval_dataloader (DataLoader | DataSpec | Evaluator | Sequence[Evaluator], optional): The :class:`.DataLoader`,
+        eval_dataloader (Iterable | DataSpec | Evaluator | Sequence[Evaluator], optional): The :class:`.Iterable`,
```
Comment (dakinggg) [inline composer/trainer/trainer.py:472]: looks like this pr includes commits from your other pr :)

File: tests/loggers/test_console_logger.py [hunk 4]
```diff
@@ -287,3 +288,49 @@ def test_console_logger_with_a_callback(console_logger_test_stream, console_logg
     expected_speed_monitor_lines = num_speed_monitor_lines_per_log_event * expected_num_logging_events
 
     assert actual_num_speed_monitor_lines == expected_speed_monitor_lines
+
+
+@pytest.mark.filterwarnings('ignore:Cannot split tensor of length .* into batches of size .*:UserWarning')
+def test_console_logger_overlapping(console_logger_test_stream, console_logger_test_file_path):
+
+    batch_size = 1
+    dataset_size = 96
```
Comment (dakinggg) [inline tests/loggers/test_console_logger.py:297]: ideally we would make the dataset size smaller so this test is faster. Does the dataset need to be this large for some reason?

File: tests/loggers/test_console_logger.py [hunk 5]
```diff
@@ -287,3 +288,49 @@ def test_console_logger_with_a_callback(console_logger_test_stream, console_logg
     expected_speed_monitor_lines = num_speed_monitor_lines_per_log_event * expected_num_logging_events
 
     assert actual_num_speed_monitor_lines == expected_speed_monitor_lines
+
+
+@pytest.mark.filterwarnings('ignore:Cannot split tensor of length .* into batches of size .*:UserWarning')
+def test_console_logger_overlapping(console_logger_test_stream, console_logger_test_file_path):
```
Comment (dakinggg) [inline tests/loggers/test_console_logger.py:294]: could you add a comment explaining what this test is testing? I'm not actually sure what is testing/why the optimizermonitor is involved

**Comment by rishab-partha** [comment]
Moving this onto new PR with new branch.

**Pull request closed without merging**